projects/vdk-plugins/vdk-impala/src/vdk/plugin/impala/ingest_to_impala.py

```diff
@@ -0,0 +1,87 @@
+# Copyright 2021 VMware, Inc.
+# SPDX-License-Identifier: Apache-2.0
+import logging
+from typing import List
+from typing import Optional
+from typing import Tuple
+
+from vdk.internal.builtin_plugins.connection.pep249.interfaces import PEP249Cursor
+from vdk.internal.builtin_plugins.ingestion.ingester_base import IIngesterPlugin
+from vdk.internal.builtin_plugins.run.job_context import JobContext
+from vdk.internal.core import errors
+
+_log = logging.getLogger(__name__)
+
+
+class IngestToImpala(IIngesterPlugin):
+    def __init__(self, context: JobContext):
+        self._context = context
+
+    def ingest_payload(
+        self,
+        payload: List[dict],
+        destination_table: Optional[str],
+        target: Optional[str] = None,
+        collection_id: Optional[str] = None,
+    ) -> None:
+        """
+        See parent class doc for details
+        """
+
+        _log.info(
+            f"Ingesting payloads to table: {destination_table} in Impala database; "
+            f"collection_id: {collection_id}"
+        )
+
+        with self._context.connections.open_connection(
+            "IMPALA"
+        ).connect() as connection:
+            cursor = connection.cursor()
+            query, parameters = self._populate_query_parameters_tuple(
+                destination_table, cursor, payload
+            )
+
+            try:
+                cursor.execute(query, parameters)
+                connection.commit()
+                _log.debug("Payload was ingested.")
+            except Exception as e:
+                errors.log_and_rethrow(
+                    errors.find_whom_to_blame_from_exception(e),
+                    _log,
+                    "Failed to send payload",
+                    "Unknown error. Error message was : " + str(e),
+                    "Will not be able to send the payload for ingestion",
+                    "See error message for help ",
+                    e,
+                    wrap_in_vdk_error=True,
+                )
+
+    @staticmethod
+    def _populate_query_parameters_tuple(
+        destination_table: str, cursor: PEP249Cursor, payload: List[dict]
+    ) -> (str, Tuple[str]):
+        """
+        Returns insert into destination table tuple of query and parameters;
+        E.g. for a table dest_table with columns val1, val2 and payload size 3, this method will return:
+        'INSERT INTO dest_table (val1, val2) VALUES (%s, %s), (%s, %s), (%s, %s)', ['val1', 'val2']
```

Comment:
nit: Method signature says `(str, Tuple[str])` is returned. So maybe, the end of this line should be `('val1', 'val2')`, instead of `['val1', 'val2']`?

Also, is there a specific reason for the column names to be `val1` and `val2`? I am left with the impression that val1 and val2 after `dest_table` are the same as the 'val1' and 'val2' at the end of the line.

Reply:
They are the same.

Reply:
Ok, sorry, I cannot express myself properly.
So, how does this example correspond to what is actually returned by the method? From the return tuple, the last line, `tuple(obj[column] for obj in payload for column in columns)` returns the cell values, not column names.
For example, if we have a payload, `[{"col1": "val1", "col2": "val2"}, {"col1": "val3", "col2": "val4"}]`, the line mentioned above would return `("val1", "val2", "val3", "val4")`, and not `("col1", "col2")`.